text/3716-target-modifiers.md

```diff
@@ -0,0 +1,514 @@
+  of target modifiers.
+* As a future extension we may be able to relax the rules to allow some
+  specific kinds of mismatches.
+* This RFC does not stabilize any target modifiers. That should happen in
+  follow-up MCPs/FCPs/RFCs/etc.
+
+# Motivation
+[motivation]: #motivation
+
+As Rust expands into low-level domains, there will be a need for precise
+control over how code is compiled. This often manifests as a new compiler flag.
+Some of these flags trigger undefined behavior if used incorrectly, which is in
+tension with Rust's safety goals. This RFC proposes a new mechanism to allow
+use of such flags while also preventing undefined behavior.
+
+The primary goal of this RFC is to unblock *stabilization* of target modifier
+flags. Adding them as unstable (and unsound) flags is already happening today
+without this RFC.
+
+## The Linux Kernel
+
+The Linux Kernel has run into a handful of cases where it is necessary to tweak
+the ABI used in the kernel. Often, this is done conditionally depending on a
+configuration option. A few examples:
+
+* When using `CONFIG_SHADOW_CALLSTACK` the x18 register must be reserved in the
+  ABI with `-Zfixed-x18`.
+* The `-Ctarget-feature=-neon` flag is used to prevent use of floating points
+  on arm.
+* On 32-bit x86, `-Zreg-struct-return` and `-Zregparm=3` are used.
+* When using `CONFIG_CFI_CLANG` the kCFI sanitizer is enabled with
+  `-Zsanitizer=kcfi`. Unlike most other sanitizers, this sanitizer is used in
+  production.
+* In several different cases, `-Zpatchable-function-entry` is used to add nops
+  before or after the function entrypoint. When mixed with `-Zsanitizer=kcfi`
+  this causes special considerations as kcfi works by placing a tag before the
+  function entrypoint.
+* To support `MITIGATION_RETPOLINE` and `MITIGATION_SLS`, `target.json` is used
+  on x86.
+
+We expect there to be more examples in the future.
+
+## Sanitizers
+
+There is [an ongoing effort to stabilize some of the sanitizers][issue123615].
+However, this effort explicitly aims to stabilize sanitizers that can be used
+without rebuilding the stdlib. With this RFC, that is no longer a blocker as
+the remaining sanitizers can be classified as a target modifier and then
+stabilized.
+
+[issue123615]: https://github.com/rust-lang/rust/issues/123615
+
+## Embedded Targets
+
+Currently, embedded platforms such as `thumb*` or `rv*` use separate targets
+for configuration with significant ABI changes. For `thumb*` targets, this is
+currently limited to Hard- vs Soft-Float which can cause issues when linked.
+However for RISC-V targets, the F (32-bit hardware float), D (64-bit hardware
+float), and Q (128-bit hardware float) extensions all can [potentially change
+the ABI][riscv-float], which would increase the number of required targets. The
+[E extension][riscv-e] may also change the ABI by limiting the number of
+registers used by the I (integer operations) extension.
```

Comment:
Actually RISC-V is one of the few targets that handles this properly. The target features control what instructions codegen is allowed to use while the ABI is specified [separately](https://github.com/rust-lang/rust/blob/f61306d47bc98af8bb9d15f1adf6086785590a8c/compiler/rustc_target/src/spec/targets/riscv64imac_unknown_none_elf.rs#L25) in the target spec. This means that you can enable FP instructions while keeping a soft-float ABI. Notice how the E target uses a [separate ABI](https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/targets/riscv32e_unknown_none_elf.rs#L23).

It's the other targets that often have target features which change the ABI.

Reply:
The change that does that was merged in the time between writing and posting this RFC.

Reply:
I know, because I'm the one that signed off on that change being merged.

And rv32e target **HAS** to have its ilp32e ABI specified, because that's the only avenue for how you inform LLVM that you want to do codegen for rv32e.

Reply:
> It's the other targets that often have target features which change the ABI.

So, that's what I thought, too. That RISC-V extensions were additive.

How do these "subtractive" extension play with target triples that define a base set of extensions, e.g. is this sound

```
rustc --target riscv32imac-unknown-none-elf --target-features e
```

Would it make sense to add a bare set of `riscv[32|64]-unknown-<platform>-<libc>` target triples (in addition to current triples), and allow user to explicitly add all of the extensions they want to use?

Reply:
> And rv32e target **HAS** to have its ilp32e ABI specified, because that's the only avenue for how you inform LLVM that you want to do codegen for rv32e.

It looks like the `RV32E` extension is incompatible with the `D` extension because of the `ILP32E` calling convention (from [riscv-elf-psabi-doc](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#ilp32e-calling-convention)):

```
The ILP32E calling convention is not compatible with ISAs that have registers that require load and store alignments of more than 32 bits.

In particular, this calling convention must not be used with the D ISA extension.
```

Reply:
> So, that's what I thought, too. That RISC-V extensions were additive.

If no abiname is specified, LLVM hits the path which infers the name based on the target feature set, which causes a mutation of the ABI: values that were once passed via stack or integer registers then get passed via the float registers.

Reply:
I don't know the riscv situation so well. Happy to reword if the paragraph is not accurate.

cc @jamesmunns @rmsyn @9names

Reply:
> How do these "subtractive" extension play with target triples that define a base set of extensions, e.g. is this sound
> 
> ```
> rustc --target riscv32imac-unknown-none-elf --target-features e
> ```

E is not an extension, RV32E and RV32I are base ISA's.
I would think that an attempt to use `--target-features e` on an I target should be rejected.

Reply:
> I don't know the riscv situation so well. Happy to reword if the paragraph is not accurate.

So, it looks accurate to me. Maybe RV32E is related, but somewhat tangential to this RFC. It may be something that requires its own RFC: when to introduce a new set of target triples for RISC-V targets.

It's a situation that should be fairly rare, especially as the ISA matures even more. However, there are no target triples that specify a `riscv32e-unknown-none-elf`, for example.

I'm open to working on such an RFC, and linking back to this RFC for the discussion of incompatible RISC-V extensions.

The more I think about it, the more adding a base set of bare RISC-V targets makes sense (following from the `riscv64-linux-android` target):

- `riscv32-unknown-none-elf`
- `riscv64-unknown-none-elf`
- ...

> I would think that an attempt to use --target-features e on an I target should be rejected.

I agree, that would make sense to me as well. Unless LLVM is doing some magic to resolve when both `E` and `I` extensions are requested, only `E` is used?

Even then, it would be nice for `rustc` to reject the pairing using the techniques discussed in this RFC.

Reply:
> I don't know the riscv situation so well. Happy to reword if the paragraph is not accurate.

So, it looks accurate to me. Maybe the discussion about target triples is related, but somewhat tangential to this RFC. It may be something that requires its own RFC: when to introduce a new set of target triples for RISC-V targets.

It's a situation that should be fairly rare, especially as the ISA matures even more. However, there are no target triples that specify a `riscv32e-unknown-none-elf`, for example.

I'm open to working on such an RFC, and linking back to this RFC for the discussion of incompatible RISC-V extensions.

The more I think about it, the more adding a base set of bare RISC-V targets makes sense (following from the `riscv64-linux-android` target):

- `riscv32-unknown-none-elf`
- `riscv64-unknown-none-elf`
- ...

> I would think that an attempt to use --target-features e on an I target should be rejected.

I agree, that would make sense to me as well. Unless LLVM is doing some magic to resolve when both `E` and `I` extensions are requested, only `E` is used?

Even then, it would be nice for `rustc` to reject the pairing using the techniques discussed in this RFC.

More rejections for `--target-features e`:

- `riscv32gc-*`
- `riscv64gc-*`
- any other targets that implicitly include the `D` and/or `I` extension
- `--target-features d`
- `--target-features i`

Reply:
LLVM rejects it.
```c++
  if (HasI && HasE)
    return getIncompatibleError("i", "e");
```

Reply:
@rmsyn 
> It's a situation that should be fairly rare, especially as the ISA matures even more. However, there are no target triples that specify a riscv32e-unknown-none-elf, for example.

Are you sure about that? https://github.com/rust-lang/rust/blob/a06b7cbe21967a86050fa92dab843c8afda1c28e/compiler/rustc_target/src/spec/targets/riscv32e_unknown_none_elf.rs

Reply:
> @rmsyn
> 
> > It's a situation that should be fairly rare, especially as the ISA matures even more. However, there are no target triples that specify a riscv32e-unknown-none-elf, for example.
> 
> Are you sure about that? https://github.com/rust-lang/rust/blob/a06b7cbe21967a86050fa92dab843c8afda1c28e/compiler/rustc_target/src/spec/targets/riscv32e_unknown_none_elf.rs

Ah, I guess not... I was looking at the output of `rustc --print target-list` on v1.82.0 stable. Thanks for the link.

Reply:
>  This means that you can enable FP instructions while keeping a soft-float ABI. 

That's actually fine on most targets, e.g. on x86 if you set the `soft-float` target feature then you can enabled x87 and SSE and whatnot and keep using the soft-float ABI.

The interesting question is what happens on RISC-V when I *disable* the FP instructions on a target that usually uses the hardfloat ABI. Sadly, what LLVM usually does in that case is silently fall back to the softfloat ABI. But I haven't checked for RISC-V.